File: python/53-maximum-subarray.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        "DP, Time: O(n), Space: O(n)"
        "dp[i]: max sum of sub-array from 0 to i include i"
        "dp[i] = max(dp[i-1]+nums[i], nums[i])"
        "dp[0] = nums[0]"
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        for i in range(1, n):
            dp[i] = max(dp[i-1]+nums[i], nums[i])
        return max(dp)
    # def maxSubArray(self, nums: List[int]) -> int:
    #     "DP, Time: O(n), Space: O(1)"
    #     n = len(nums)
    #     pre = nums[0]
    #     cur = 0
    #     result = pre
    #     for i in range(1, n):
    #         cur = max(pre+nums[i], nums[i])
    #         result = max(result, cur)
    #         pre = cur
    #     return result

nums = [-2,1,-3,4,-1,2,1,-5,4]
result = Solution().maxSubArray(nums)
assert result == 6

nums = [1]
result = Solution().maxSubArray(nums)
assert result == 1

nums = [0]
result = Solution().maxSubArray(nums)
assert result == 0

nums = [-1]
result = Solution().maxSubArray(nums)
assert result == -1

nums = [-2147483647]
result = Solution().maxSubArray(nums)
assert result == -2147483647

File: cpp/53. Maximum Subarray.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        ll prefixsum = 0; // sum in [0, i]
        ll curmin = 0; // min of all prefixsum in [0, 0], [0, 1], [0, 2],..., [0, i-1]
        // prefixsum - curmin = max subarray sum end at i
        ll ans = -1e18;
        for(int i = 0; i < n; ++i) {
            prefixsum += nums[i];
            ans = max(ans, prefixsum - curmin);
            curmin = min(curmin, prefixsum);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; int ans;

    solution = Solution();
    nums = {-2,1,-3,4,-1,2,1,-5,4};
    ans = solution.maxSubArray(nums);
    assert(ans==6);

    solution = Solution();
    nums = {1};
    ans = solution.maxSubArray(nums);
    assert(ans==1);

    solution = Solution();
    nums = {0};
    ans = solution.maxSubArray(nums);
    assert(ans==0);

    solution = Solution();
    nums = {-1};
    ans = solution.maxSubArray(nums);
    assert(ans==-1);

    solution = Solution();
    nums = {-2147483647};
    ans = solution.maxSubArray(nums);
    assert(ans==-2147483647);
}